src/Collections/MemoryConfigCollection.php

```diff
@@ -0,0 +1,306 @@
+     */
+    protected $callCache = [];
+
+    /**
+     * @var array
+     */
+    protected $metadata = [];
+
+    /**
+     * @var array
+     */
+    protected $history = [];
+
+    /**
+     * @var boolean
+     */
+    protected $trackMetadata = false;
+
+    /**
+     * ConfigCollection constructor.
+     *
+     * @param bool $trackMetadata
+     */
+    public function __construct($trackMetadata = false)
+    {
+        $this->trackMetadata = $trackMetadata;
+    }
+
+    /**
+     * @return static
+     */
+    public static function create()
+    {
+        return new static();
+    }
+
+    /**
+     * Trigger transformers to load into this store
+     *
+     * @param  TransformerInterface[] $transformers
+     * @return $this
+     */
+    public function transform($transformers)
+    {
+        foreach ($transformers as $transformer) {
+            $transformer->transform($this);
+        }
+        return $this;
+    }
+
+    public function set($class, $name, $data, $metadata = [])
+    {
+        $class = strtolower($class);
+        $this->saveMetadata($class, $metadata);
+
+        if ($name) {
+            if (!isset($this->config[$class])) {
+                $this->config[$class] = [];
+            }
+            $this->config[$class][$name] = $data;
+        } else {
+            $this->config[$class] = $data;
+        }
+
+        // Flush call cache for this class, and any subclasses
+        unset($this->callCache[$class]);
+        foreach ($this->callCache as $nextClass => $data) {
+            if (is_subclass_of($nextClass, $class, true)) {
+                unset($this->callCache[$nextClass]);
+            }
+        }
+        return $this;
+    }
+
+    public function get($class, $name = null, $options = 0)
+    {
+        if (!is_int($options) && $options !== true && !is_array($options)) {
+            throw new \InvalidArgumentException("Invalid config option: " . var_export($options, true));
+        }
+
+        // Get config for complete class
+        $class = strtolower($class);
+        $config = $this->getClassConfig($class, $options);
+
+        // Return either name, or whole-class config
+        if ($name) {
+            return isset($config[$name]) ? $config[$name] : null;
+        }
+        return $config;
+    }
+
+    /**
+     * @param string $class
+     * @param mixed  $options
+     * @return array|null
+     */
+    protected function getClassConfig($class, $options)
+    {
+        $class = strtolower($class);
+
+        // Can't apply middleware to config on non-existant class
+        if (!isset($this->config[$class])) {
+            return null;
+        }
+
+        // check if middleware needs applying
+        $disableFlag = isset($options['disableFlag']) ? $options['disableFlag'] : $options;
+        if ($disableFlag === true) {
+            return $this->config[$class];
+        }
+
+        // Check cache
+        if (isset($this->callCache[$class][$disableFlag])) {
```

Comment:
For the sake of simplicity, I think I should simply remove the array-form, and keep this as a simple int for flag-assignment, and rename to something that implies it's purpose more succinctly.